source/common/grpc/codec.cc

```diff
@@ -98,5 +98,71 @@ bool Decoder::decode(Buffer::Instance& input, std::vector<Frame>& output) {
   return true;
 }
 
+uint64_t FrameInspector::decode(Buffer::Instance& data) {
+  uint64_t count = data.getRawSlices(nullptr, 0);
+  STACK_ARRAY(slices, Buffer::RawSlice, count);
```

Comment:
I don't have a good heuristic what' more expensive:
1. stack arrays like this implementation.
2. using data copyOut to scan internal raw buffers.

I think CPU-wise #1 is better, but it costs memory.

Reply:
(1) is good, thanks

Reply:
Sorry, I'm late to this PR, but why isn't this code consolidated with `Decoder::decode`? This kind of non-trivial parsing, motivating https://github.com/envoyproxy/envoy/issues/8740), should ideally be only done once.

Reply:
I asked for this to happen in an earlier comment but @kyessenov said it wasn't possible for some reason. If it is possible we should unify it in a follow up.

Reply:
It would require refactoring to get all the side-effects from `decode` to be wrapped in lambdas I think. I can try to take a stab at it in a follow-up.

Reply:
@kyessenov I think a visitor pattern might work here, WDYT?

Reply:
Yes, I was thinking the same. We can make the frame inspector a base visitor that the decoder extends with the frame collection.